### PR TITLE
Add Login, SignUp, Logout functionality

### PR DESCRIPTION
**Pull request by jynti**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -3,6 +3,7 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
 ruby '2.4.1'
 
+gem 'bootstrap-sass', '~> 3.3.6'
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
```
Comment (aishmita-aggarwal) [inline Gemfile]: @jynti Please remove all the unwanted commented code from across the application.

File: Gemfile [hunk 2]
```diff
@@ -60,3 +59,4 @@ end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+gem 'jquery-rails'
```
Comment (aishmita-aggarwal) [inline Gemfile]: @jynti Please make sure, all gems have version specifications.

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -10,7 +10,12 @@
...
 //
-//= require rails-ujs
+//= require jquery3
+//= require jquery_ujs
 //= require activestorage
-//= require turbolinks
 //= require_tree .
+//= require bootstrap-sprockets
+
+$( document ).ready(function() {
```
Comment (aishmita-aggarwal) [inline app/assets/javascripts/application.js]: @jynti Do we need this? If not, please remove!

File: app/assets/javascripts/sessions.coffee [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (aishmita-aggarwal) [inline app/assets/javascripts/sessions.coffee]: @jynti Please remove any file which is not producing any code!

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,2 +1,8 @@
 class ApplicationController < ActionController::Base
+  helper_method :current_user
+
+  def current_user
+    @current_user ||= User.find(session[:user_id]) if session[:user_id]
```
Comment (aishmita-aggarwal) [inline app/controllers/application_controller.rb]: @jynti Why have we used find and not find_by here?

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
+class SessionsController < ApplicationController
+
+  before_action :ban_user, only: :new
```
Comment (aishmita-aggarwal) [inline app/controllers/sessions_controller.rb]: @jynti Should we not have this action before create as well?

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:email])
```
Comment (aishmita-aggarwal) [inline app/controllers/sessions_controller.rb]: @jynti Please move data loading logic to a before_action.

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      flash[:success] = 'Logged In!'
```
Comment (aishmita-aggarwal) [inline app/controllers/sessions_controller.rb]: @jynti Please use i18n wherever possible.

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def destroy
+    session[:user_id] = nil
+    flash[:success] = "Logged out!"
+    redirect_to root_url
+  end
+
+
+  def ban_user
```
Comment (aishmita-aggarwal) [inline app/controllers/sessions_controller.rb]: @jynti Can we find a better name for this method? Also, should it be private/public or protected?

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      flash[:success] = 'Thank you for signing up!'
+      redirect_to root_url
+    else
+      flash.now[:info] = 'Try again'
+      render :new
+    end
+  end
+
+  private
+  def allowed_params
```
Comment (aishmita-aggarwal) [inline app/controllers/users_controller.rb]: @jynti  Please fix indentation! Also, please use appropriate method names.

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+class User < ApplicationRecord
+  has_secure_password
+  validates :fname, :lname, :email, presence: true
```
Comment (aishmita-aggarwal) [inline app/models/user.rb]: @jynti Please add the following validations:

1. fname, lname should not be of more than 255 characters length.
2. email should be unique.
3. password and password_confirmation presence. Ignore if already present.

File: app/views/users/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,56 @@
+<h1>Sign Up</h1>
+
+<%= form_for @user, html: {class: 'form-horizontal'} do |f| %>
+  <% if @user.errors.any? %>
```
Comment (aishmita-aggarwal) [inline app/views/users/new.html.erb]: @jynti Please remove this.

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    <div class="error_messages">
+      <h4>Form is invalid</h4>
+    </div>
+  <% end %>
+
+  <div class="field form-group">
+    <%= f.label :fname, class: 'col-sm-2 control-label' %>
+    <div class="col-sm-5">
+    <%= f.text_field :fname, class: 'form-control' %>
+    <%= error_message_on f.object, :fname%>
```
Comment (aishmita-aggarwal) [inline app/views/users/new.html.erb]: @jynti Please fix indentation!

File: app/views/users/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    <%= f.label :password_confirmation, class: 'col-sm-2 control-label' %>
+    <div class="col-sm-5">
+    <%= f.password_field :password_confirmation, class: 'form-control' %>
+    <%= error_message_on f.object, :password_confirmation%>
+    </div>
+  </div>
+
+  <div class="actions form-group">
+    <div class="col-sm-offset-2 col-sm-10">
+      <%= f.submit 'Sign Up ', class: 'btn btn-primary' %>
```
Comment (aishmita-aggarwal) [inline app/views/users/new.html.erb]: @jynti Please fix typo

File: config/routes.rb [hunk 15]
```diff
@@ -1,3 +1,13 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root to:"welcome#index"
+  resources :users
+
+  get '/login' =>'sessions#new', as: 'login'
```
Comment (aishmita-aggarwal) [inline config/routes.rb]: @jynti Can we use better way of writing routes here?

File: db/migrate/20180228090121_create_users.rb [hunk 16]
```diff
@@ -0,0 +1,12 @@
+class CreateUsers < ActiveRecord::Migration[5.2]
+  def change
+    create_table :users do |t|
+      t.string :fname
```
Comment (aishmita-aggarwal) [inline db/migrate/20180228090121_create_users.rb:4]: @jynti Please use valid attributes names:

For example: fname should have been first_name.

File: test/controllers/sessions_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,9 @@
+require 'test_helper'
```
Comment (aishmita-aggarwal) [inline test/controllers/sessions_controller_test.rb]: @jynti Please remove all test files. And do not use rails generators for models and controllers.

File: app/controllers/sessions_controller.rb [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+  private
+    def prevent_current_user_from_login
+      if current_user
+        flash[:warning] = t('sessions.prevent_login')
+        redirect_to root_url
+      end
+    end
+
+    def find_current_user
```
Comment (aishmita-aggarwal) [inline app/controllers/sessions_controller.rb]: @jynti Are we finding the current user through this method?

File: app/controllers/sessions_controller.rb [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  def destroy
+    session[:user_id] = nil
+    flash[:success] = t('.logged_out')
+    redirect_to root_url
+  end
+
+  private
+    def prevent_current_user_from_login
```
Comment (aishmita-aggarwal) [inline app/controllers/sessions_controller.rb]: @jynti Please rename this to 'redirect_when_logged_in'.

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+      render :new
+    end
+  end
+
+  private
+    def user_params
+      params.require(:user).permit(:first_name, :last_name, :email, :password, :password_confirmation)
+    end
+
+    def prevent_current_user_from_signup
```
Comment (aishmita-aggarwal) [inline app/controllers/users_controller.rb]: @jynti Please rename this to redirect_when_logged_in

File: app/views/layouts/application.html.erb [hunk 21]
```diff
@@ -8,7 +8,30 @@
...
-    <%= yield %>
-  </body>
+  <nav class="navbar navbar-inverse">
+    <div class="container-fluid">
+      <div class="navbar-header">
+        <%= link_to 'Flash Site', root_url, class: "navbar-brand" %>
+      </div>
+      <div class="nav navbar-nav navbar-right">
+        <% if current_user %>
+           <strong id= "user_name"><%= current_user.first_name.capitalize %></strong> <%= link_to "Logout", logout_path%>
```
Comment (aishmita-aggarwal) [inline app/views/layouts/application.html.erb]: @jynti Please make sure there are no indentation issues in the code.

File: config/locales/en.yml [hunk 22]
```diff
@@ -1,33 +1,18 @@
...
-# available at http://guides.rubyonrails.org/i18n.html.
-
 en:
-  hello: "Hello world"
+  login: 'Log In!'
+  signup: 'Sign Up!'
+
+  users:
+    create:
+      success_signup: 'Thank you for signing up!'
```
Comment (aishmita-aggarwal) [inline config/locales/en.yml:7]: @jynti Let's use the default flow of i18n messages. Let's discuss this.

File: test/fixtures/users.yml [hunk 23]
```diff
@@ -0,0 +1,13 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (aishmita-aggarwal) [inline test/fixtures/users.yml]: @jynti Do we need this file?

File: app/views/layouts/application.html.erb (hunk 21 above)
Comment (jynti) [reply app/views/layouts/application.html.erb]: Done.

File: config/locales/en.yml (hunk 22 above)
Comment (jynti) [reply config/locales/en.yml:7]: Done

**Pull request merged**